Refactor PEMethodSymbol loading of UnmanagedCallersOnly

333fred opened the pull request:
My simple implementation of GetUnmanagedCallersOnlyAttributeData in PEMethodSymbol would cause all PEMethodSymbols to load all attributes when the data is requested, unnecessarily increasing memory pressure. This refactors the implementation to only load the one attribute we need for it, and to not load the public-facing information that is unneeded for this. In the process, I refactored the method checks to share between regular SourceMethodSymbolWithAttributes and PEMethodSymbols, so the criteria for should be more consistent.

333fred commented:
@dotnet/roslyn-compiler @AlekseyTs for review please.

AlekseyTs commented on src/Compilers/Core/Portable/MetadataReader/PEModule.cs:
```diff
@@ -1134,6 +1134,47 @@ internal bool HasIsByRefLikeAttribute(EntityHandle token)
+                    var numNamed = sigReader.ReadUInt16();
+                    for (int i = 0; i < numNamed; i++)
+                    {
+                        var ((name, value), _, _) = attributeArgumentDecoder.DecodeCustomAttributeNamedArgumentOrThrow(ref sigReader);
```
>_, _ [](start = 44, length = 4)

Why are we ignoring these values?

333fred replied:
The first one is whether this is a property or not, I can check that. The second is the typeCode of the deserialized data, and that is checked by the main decode step itself.

---
In reply to: [493008750](https://github.com/dotnet/roslyn/pull/47950#discussion_r493008750) [](ancestors = 493008750)

AlekseyTs replied:
> The second is the typeCode of the deserialized data, and that is checked by the main decode step itself.

Consider adding a comment about this.

---
In reply to: [493888966](https://github.com/dotnet/roslyn/pull/47950#discussion_r493888966) [](ancestors = 493888966,493008750)

333fred replied:
Now actually checked. I added tests that will fail without these.

---
In reply to: [493918626](https://github.com/dotnet/roslyn/pull/47950#discussion_r493918626) [](ancestors = 493918626,493888966,493008750)

AlekseyTs commented on src/Compilers/Core/Portable/MetadataReader/PEModule.cs:
```diff
@@ -1134,6 +1134,47 @@ internal bool HasIsByRefLikeAttribute(EntityHandle token)
+                        var namedArgumentDecoded = unmanagedCallersOnlyDecoder(name, value);
+                        if (namedArgumentDecoded.IsCallConvs)
+                        {
+                            isValid = isValid && namedArgumentDecoded.IsValid;
```
>isValid && [](start = 38, length = 10)

Why combine validity state with previous values when we are only interested in the last name/value pair?

AlekseyTs commented on src/Compilers/Core/Portable/MetadataReader/PEModule.cs:
```diff
@@ -1134,6 +1134,47 @@ internal bool HasIsByRefLikeAttribute(EntityHandle token)
+                }
+                catch (Exception ex) when (ex is BadImageFormatException or UnsupportedSignatureContent)
+                {
+                    return UnmanagedCallersOnlyAttributeData.Create(ImmutableHashSet<INamedTypeSymbolInternal>.Empty, isValid: false);
```
>return UnmanagedCallersOnlyAttributeData.Create(ImmutableHashSet<INamedTypeSymbolInternal>.Empty, isValid: false) [](start = 20, length = 113)

It looks like we simply ignore presence of other well-known attributes if we cannot crack them due to an invalid metadata.

333fred replied:
I'm not sure that's true. At least when cracking `Obsolete`, we're inconsistent. We appear to have this behavior:
* If the constructor is either `public ObsoleteAttribute(string? message)` or `public ObsoleteAttribute(string? message, bool error)` and there is bad metadata in the constructor itself, we ignore the attribute entirely.
* If the constructor is perfectly fine, and we encounter bad metadata cracking named properties, we treat the attribute as having whatever the values it has from the constructor, and we default the DiagnosticId and UrlFormat.

IMO, it's better to replicate this second behavior with this attribute, rather than the first behavior.

---
In reply to: [493012289](https://github.com/dotnet/roslyn/pull/47950#discussion_r493012289) [](ancestors = 493012289)

AlekseyTs replied:
> IMO, it's better to replicate this second behavior with this attribute, rather than the first behavior.

A well-known attribute is matched by a constructor signature, if we cannot crack the signature, or it doesn't match something we recognize, we should ignore the attribute.

---
In reply to: [493892346](https://github.com/dotnet/roslyn/pull/47950#discussion_r493892346) [](ancestors = 493892346,493012289)

333fred replied:
> A well-known attribute is matched by a constructor signature

If we're in this part of decoding, we matched the constructor signature already, and are going through property/field assignments. If we didn't match to the known constructor signature, we return `null` above, ignoring the attribute.

---
In reply to: [493916803](https://github.com/dotnet/roslyn/pull/47950#discussion_r493916803) [](ancestors = 493916803,493892346,493012289)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs:
```diff
@@ -967,45 +968,105 @@ protected static UnmanagedCallersOnlyAttributeData DecodeUnmanagedCallersOnlyAtt
+                    if (namedArgumentDecoded.IsCallConvs)
                     {
-                        isValid = false;
+                        isValid = isValid && namedArgumentDecoded.IsValid;
```
>isValid [](start = 34, length = 7)

Since we are interested only in the last value, why are we paying attention to validity of other values?

333fred replied:
> Since we are interested only in the last value

If we're interpreting source, I don't believe it's true that we're only interested in the last thing. I use the set of the last convention as the set for simplicity, but the data in the UnmanagedCallersOnlyAttribute itself is definitively bad as you can't specify it more than once. I can likely remove this step from the IL decoding version though.

---
In reply to: [493019870](https://github.com/dotnet/roslyn/pull/47950#discussion_r493019870) [](ancestors = 493019870)

333fred replied:
I'm actually going to remove the `IsValid` tracking entirely. We only use it for use-site diagnostics on imported method symbols, and I don't think it's worth the amount of effort I'm putting in to track it. If we get feedback from users about wanting better error messages from bad metadata, we can reconsider.

---
In reply to: [493882174](https://github.com/dotnet/roslyn/pull/47950#discussion_r493882174) [](ancestors = 493882174,493019870)

AlekseyTs replied:
> If we're interpreting source, I don't believe it's true that we're only interested in the last thing.

We should be consistent even when we are recovering from errors. We recognize properties on Obsolete attribute, let's match that behavior in terms which value to use as a result of decoding in presence of duplicate named value.

---
In reply to: [493894825](https://github.com/dotnet/roslyn/pull/47950#discussion_r493894825) [](ancestors = 493894825,493882174,493019870)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs:
```diff
@@ -967,45 +968,105 @@ protected static UnmanagedCallersOnlyAttributeData DecodeUnmanagedCallersOnlyAtt
-                    }
-                    callingConventionTypes = builder.ToImmutableHashSet();
-                    builder.Free();
+            if (!UnmanagedCallersOnlyAttributeData.IsCallConvsTypedConstant(key, in value))
```
>key [](start = 76, length = 3)

Where do we verify that this refers to a property?

333fred replied:
I need to rename this, it's actually a field. I'll update.

---
In reply to: [493035146](https://github.com/dotnet/roslyn/pull/47950#discussion_r493035146) [](ancestors = 493035146)

333fred replied:
When decoding from source, we our existing `NamedArguments` infrastructure does not keep track of whether it's a field or a property, so I cannot verify that. I can do so when decoding from IL though.

---
In reply to: [493878851](https://github.com/dotnet/roslyn/pull/47950#discussion_r493878851) [](ancestors = 493878851,493035146)

AlekseyTs replied:
> When decoding from source, we our existing NamedArguments infrastructure does not keep track of whether it's a field or a property, so I cannot verify that. I can do so when decoding from IL though.

We need to check this from source as well. See how we are doing similar check for Obsolete attribute in AttributeData.DecodeObsoleteAttribute

---
In reply to: [493880852](https://github.com/dotnet/roslyn/pull/47950#discussion_r493880852) [](ancestors = 493880852,493878851,493035146)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs:
```diff
@@ -884,56 +884,34 @@ private void DecodeModuleInitializerAttribute(DecodeWellKnownAttributeArguments<
-            var (returnTypeSyntax, genericTypeParameters) = SyntaxNode switch
+            bool reportedError = CheckAndReportValidUnmanagedCallersOnlyTarget(arguments.AttributeSyntaxOpt.Name.Location, arguments.Diagnostics);
+
+            var returnTypeSyntax = SyntaxNode switch
```
>SyntaxNode switch [](start = 35, length = 17)

It looks like we have an existing helper for that, ```ExtractReturnTypeSyntax```. Let's not invent new ways to do the same thing.

AlekseyTs commented:
Done with review pass (iteration 2)

AlekseyTs commented on src/Compilers/Core/Portable/MetadataReader/PEModule.cs:
```diff
@@ -1134,6 +1134,46 @@ internal bool HasIsByRefLikeAttribute(EntityHandle token)
+                        if (namedArgumentDecoded.IsCallConvs)
+                        {
+                            unmanagedConventionTypes = namedArgumentDecoded.CallConvs;
+                            break;
```
>break; [](start = 28, length = 6)

Is bailing out on the first value consistent with what we do for Obsolete attribute properties?

AlekseyTs replied:
See CrackObsoleteProperties

---
In reply to: [493960320](https://github.com/dotnet/roslyn/pull/47950#discussion_r493960320) [](ancestors = 493960320)

333fred replied:
Yes. The decoder there takes the first non-null value for each property it is interested in.

---
In reply to: [493961349](https://github.com/dotnet/roslyn/pull/47950#discussion_r493961349) [](ancestors = 493961349,493960320)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs:
```diff
@@ -958,54 +959,111 @@ internal bool CalculateUseSiteDiagnostic(ref DiagnosticInfo result)
+                    // Technically, CIL can define a field and a property with the same name. However, such a
+                    // member results in an Ambiguous Member error, and we never get to piece of code at all.
+                    // See UnmanagedCallersOnly_PropertyAndFieldNamedCallConvs for an example
+                    bool isField = attribute.AttributeClass.GetMembers(key).Any(m => m is FieldSymbol);
```
>m => m is FieldSymbol [](start = 80, length = 21)

It loos like fo Obsolete attribute properties we also check their type.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEMethodSymbol.cs:
```diff
@@ -1439,24 +1437,15 @@ internal override ObsoleteAttributeData ObsoleteAttributeData
-                var allAttributes = GetAttributes();
+                var containingModule = (PEModuleSymbol)ContainingModule;
+                var unmanagedCallersOnlyData = containingModule.Module.TryGetUnmanagedCallersOnlyAttribute(_handle, new MetadataDecoder(containingModule),
+                    static (name, value, isProperty) => MethodSymbol.TryDecodeUnmanagedCallersOnlyCallConvsField(name, value, isProperty, location: null, diagnostics: null));
```
>isProperty [](start = 41, length = 10)

It looks like PEModule passes ```!isProperty```. Rename?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs:
```diff
@@ -884,33 +884,21 @@ private void DecodeModuleInitializerAttribute(DecodeWellKnownAttributeArguments<
+            var returnTypeSyntax = this.ExtractReturnTypeSyntax();
 
-            if (isGenericMethod(this) || ContainingType.IsGenericType)
+            if (returnTypeSyntax is CompilationUnitSyntax && this is not SynthesizedSimpleProgramEntryPointSymbol)
```
>returnTypeSyntax is CompilationUnitSyntax && this is not SynthesizedSimpleProgramEntryPointSymbol [](start = 16, length = 97)

Perhaps we should simply check if what we got back is ```CSharpSyntaxTree.Dummy.GetRoot()```? The current kind check and the SynthesizedSimpleProgramEntryPointSymbol check look too cryptic.

AlekseyTs commented:
Done with review pass (iteration 2)

333fred commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs:
```diff
@@ -956,6 +930,35 @@ static void checkAndReportManagedTypes(TypeSymbol type, SyntaxNode syntax, bool
+                ImmutableHashSet<CodeAnalysis.Symbols.INamedTypeSymbolInternal>? callingConventionTypes = null;
+                if (attribute.CommonNamedArguments is { IsDefaultOrEmpty: false } namedArgs)
+                {
+                    var systemType = @this.DeclaringCompilation.GetWellKnownType(WellKnownType.System_Type);
```
This is mostly a copy/paste to move the function from `MethodSymbol` to where it actually matters. The only addition is the type check below, and retrieving `System.Type` here.

333fred commented:
@AlekseyTs addressed feedback, thanks.

---
In reply to: [698038641](https://github.com/dotnet/roslyn/pull/47950#issuecomment-698038641) [](ancestors = 698038641)

AlekseyTs commented on src/Compilers/Core/Portable/MetadataReader/PEModule.cs:
```diff
@@ -1134,6 +1134,51 @@ internal bool HasIsByRefLikeAttribute(EntityHandle token)
+                    var numNamed = sigReader.ReadUInt16();
+                    for (int i = 0; i < numNamed; i++)
+                    {
+                        // typeCode of the value is checked by the decoder itself
```
>// typeCode of the value is checked by the decoder itself [](start = 24, length = 57)

Is this comment still accurate?

AlekseyTs reviewed (approved):
LGTM (iteration 5)

jaredpar commented on src/Compilers/Core/Portable/Symbols/Attributes/UnmanagedCallersOnlyAttributeData.cs:
```diff
@@ -13,29 +13,30 @@ namespace Microsoft.CodeAnalysis
             {
-                (null, true) or ({ IsEmpty: true }, true) => PlatformDefault,
-                _ => new UnmanagedCallersOnlyAttributeData(callingConventionTypes ?? ImmutableHashSet<INamedTypeSymbolInternal>.Empty, isValid)
+                null or { IsEmpty: true } => PlatformDefault,
```
This is visually pleasing to read.

jaredpar commented on src/Compilers/Core/Portable/Symbols/Attributes/UnmanagedCallersOnlyAttributeData.cs:
```diff
@@ -13,29 +13,30 @@ namespace Microsoft.CodeAnalysis
-                (null, true) or ({ IsEmpty: true }, true) => PlatformDefault,
-                _ => new UnmanagedCallersOnlyAttributeData(callingConventionTypes ?? ImmutableHashSet<INamedTypeSymbolInternal>.Empty, isValid)
+                null or { IsEmpty: true } => PlatformDefault,
+                _ => new UnmanagedCallersOnlyAttributeData(callingConventionTypes ?? ImmutableHashSet<INamedTypeSymbolInternal>.Empty)
```
Nit: Could go full pattern matching here and do 
```suggestion
                null => new UnmanagedCallersOnlyAttributeData(ImmutableHashSet<INamedTypeSymbolInternal>.Empty),
                _ => new UnmanagedCallersOnlyAttributeData(callingConventionTypes)
```

333fred replied:
Actually, you've pointed out that the ?? is unnecessary, because `null` was already filtered out :)

jaredpar commented on src/Compilers/Core/Portable/MetadataReader/PEModule.cs:
```diff
@@ -1134,6 +1134,50 @@ internal bool HasIsByRefLikeAttribute(EntityHandle token)
         }
 
+#nullable enable
+        internal UnmanagedCallersOnlyAttributeData? TryGetUnmanagedCallersOnlyAttribute(
```
Consider adding a comment noting the importance of only loading the single attribute here.
